src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java

```diff
@@ -0,0 +1,136 @@
+package com.slack.kaldb.metadata.core;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Function;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import com.slack.kaldb.metadata.zookeeper.NodeExistsException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import org.apache.curator.utils.ZKPaths;
+import org.slf4j.Logger;
+
+/**
+ * KaldbMetadataStore is an abstract class on top of which all the metadata stores will be built.
+ *
+ * <p>This abstraction is needed so that we can limit the ZK access to the application to a specific
+ * set of paths and also to limit the operations that can be performed on those nodes. For example,
+ * we only want the SnapshotMetadata to ever be created or deleted but never updated.
+ */
+abstract class KaldbMetadataStore<T extends KaldbMetadata> {
+  protected final MetadataStore metadataStore;
+  protected final String storeFolder;
+  protected final MetadataSerializer<T> metadataSerializer;
+  protected final Logger logger;
+
+  public KaldbMetadataStore(
+      MetadataStore metadataStore,
+      String storeFolder,
+      MetadataSerializer<T> metadataSerializer,
+      Logger logger)
+      throws ExecutionException, InterruptedException {
+    checkNotNull(metadataStore, "MetadataStore can't be null");
+    checkState(
+        storeFolder != null && !storeFolder.isEmpty(),
+        "SnapshotStoreFolder can't be null or empty.");
+    checkNotNull(logger, "Logger can't be null or empty");
+
+    this.metadataStore = metadataStore;
+    this.storeFolder = storeFolder;
+    this.metadataSerializer = metadataSerializer;
+    this.logger = logger;
+
+    // Create the path to the store in ZK. However, since 2 different processes may create
+    // a path at the same time, ignore the exception if node already exists.
+    try {
+      metadataStore.create(storeFolder, "", true).get();
+    } catch (NodeExistsException e) {
+      // ignore exception, since node creation is idemponent.
+    }
+  }
+
+  protected String getPath(String snapshotName) {
+    return ZKPaths.makePath(storeFolder, snapshotName);
+  }
+
+  // TODO: byte arrays every where.
+  @SuppressWarnings("UnstableApiUsage")
+  public ListenableFuture<T> get(String path) {
```

Comment:
can we not use the name `get` here? 

Since `get` means a blocking call when used in a CompletableFuture / ListenableFuture context this method name could be something like `getData` ?

Reply:
changed to getNode.